### PR TITLE
Fix for DockerOperator Xcoms functionality

### LINKED ISSUE #18874: DockerOperator xcom is broken
### Apache Airflow version

2.0.2

### Operating System

ubuntu 20.04

### Versions of Apache Airflow Providers

apache-airflow-providers-amazon==1.3.0
apache-airflow-providers-celery==1.0.1
apache-airflow-providers-databricks==1.0.1
apache-airflow-providers-docker==1.2.0
apache-airflow-providers-ftp==1.0.1
apache-airflow-providers-http==1.1.1
apache-airflow-providers-imap==1.0.1
apache-airflow-providers-oracle==1.1.0
apache-airflow-providers-postgres==1.0.2
apache-airflow-providers-presto==1.0.2
apache-airflow-providers-sftp==1.2.0
apache-airflow-providers-sqlite==1.0.2
apache-airflow-providers-ssh==1.3.0
apache-airflow-providers-tableau==1.0.0

### Deployment

Virtualenv installation

### Deployment details

_No response_

### What happened

When using DockerOperator with 'do_xcom_push' set as True,
Airflow documentation states that without xcom_all also True, the default False behavior is that airflow would push only the last line from docker into xcom return_value, but this statement is False

do_xcom_push just returns the entire \ chunked part of the entire docker output as a single 'merged' string,
all the lines, as far as I can tell..

See code lines:
https://github.com/apache/airflow/blob/10023fdd65fa78033e7125d3d8103b63c127056e/airflow/providers/docker/operators/docker.py#L258

lines is the generator that cli.attach returns earlier in the code, the generator comes from docker py, see here:
https://github.com/docker/docker-py/blob/7172269b067271911a9e643ebdcdca8318f2ded3/docker/api/client.py#L418

Following the api code for docker-py, and into the generator logic (socket read and stuff..) 
we can conclude that the assumption that 'cli.attach' returns 'lines' is wrong, therefore do_xcom_push is braking the 'contract' set by the documentation of the parameter, it never actually returns the last line.

So it is probably better just to also use 'xcom_all' because that at least uses cli.logs which is a safer interface to use, until this issue is fixed, probably we need to create a matching issue in the docker-py repo, but maybe airflow should just handle the stream itself and split it into lines, returning the last line, see this example, modified from another issue on xcoms:

```python
from docker import APIClient                                                                                                                                                               
d = APIClient()                                                                                                                                                                            
c = d.create_container(image='ubuntu:20.04', name='TEST', command="""bash -c "echo 'test' && echo 'test2'" """, host_config=d.create_host_config(auto_remove=False, network_mode='bridge'))
gen=d.attach(c['Id'], stderr=True, stdout=True, stream=True)                                                                                                                               
d.start(c['Id'])                                                                                                                                                                           
print( [i for i in gen] )                                                                                                                                                                  
print([i for i in d.logs(c['Id'], stream=True, stdout=True, stderr=True)])                                                                                                                 
d.remove_container(c['Id'])
```

Thank you @nullhack
Related issues: 
https://github.com/apache/airflow/issues/15952
https://github.com/apache/airflow/issues/9164
https://github.com/apache/airflow/issues/14809 - This is a real nasty issue, I'm also having

### What you expected to happen

_No response_

### How to reproduce

_No response_

### Anything else

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

Edits:
- airflow version correction
- Code chunk formatting

### PR DESCRIPTION
**Pull request by asaf400**
closes: #18874

This PR is an attempt at fixing DockerOperator xcoms behavior, 
The fix is keeping the existing attach to docker's log stream for 'live' airflow logging for long-running docker containers with progress output, but asking for docker logs after the container exited for xcoms usage (either last line, or all lines),
thus restoring the behavior documented here: 
https://airflow.apache.org/docs/apache-airflow-providers-docker/stable/_api/airflow/providers/docker/operators/docker/index.html#airflow.providers.docker.operators.docker.DockerOperator

> xcom_all (bool) -- Push all the stdout or just the last line. The default is False (last line).

Please note, I do not know how Mocks works in the testing files, and so, I don't think I can do much about tests for this PR,
the current tests for this functionality exist and are currently (before this PR) passing,
even though they shouldn't, as described in #18874

the Tests are located here:
https://github.com/apache/airflow/blob/80b5e65a6abf0a136c5690548c5039f90dda01ab/tests/providers/docker/operators/test_docker.py#L455-L465

so in summary, the current tests have been useless for discovering this issue, and I personally don't know enough about airflow pytest/unittest setup to propose a change to make them better, any help or additional commit into this PR is Welcome

### CODE REVIEW

File: airflow/providers/docker/operators/docker.py [hunk 1]
```diff
@@ -304,21 +304,24 @@ def _run_image_with_mounts(self, target_mounts, add_tmp_variable: bool) -> Optio
...
+                if hasattr(log_chunk, 'decode'):
+                    # Note that log_chunk returned can also be byte sequences so we have to handle decode here
+                    log_chunk = log_chunk.decode('utf-8')
+                log_chunk = log_chunk.strip()
+                res_lines.append(log_chunk)
+                self.log.info(log_chunk)
             result = self.cli.wait(self.container['Id'])
+            # after container has exited, grab the entire log ignoring the chunked log stream that was used with attach
+            # self.cli.logs uses docker's /containers/{id}/logs, while self.cli.attach uses /containers/{id}/attach
+            lines = self.cli.logs(container=self.container['Id'], stdout=True, stderr=True, stream=True)
```
Comment (uranusjr) [inline airflow/providers/docker/operators/docker.py:324]: I feel we should ultilise the `tail` argument to fetch less data when only the last line if needed. Or even not do this additional call at all if `retrieve_output` is set and the container exited successfully.

The `_get_return_value_from_logs` can be removed entirely (it is only used once here), and it's "private" so we are allowed to break it (or we can always bump the major version of the Docker provider if needed).

Comment (asaf400) [reply airflow/providers/docker/operators/docker.py:324]: Yeah, tail looks great, I didn't know it was an argument of logs(), even though it was right there in the docs..

Comment (asaf400) [reply airflow/providers/docker/operators/docker.py:324]:
```suggestion

```

File: airflow/providers/docker/operators/docker.py [hunk 2]
```diff
@@ -328,7 +331,7 @@ def _run_image_with_mounts(self, target_mounts, add_tmp_variable: bool) -> Optio
             if self.retrieve_output:
                 ret = return_value
             elif self.do_xcom_push:
-                ret = self._get_return_value_from_logs(res_lines, line)
+                ret = self._get_return_value_from_logs(lines, lines[-1])
             return ret
```
Comment (asaf400) [inline airflow/providers/docker/operators/docker.py:335]: @uranusjr 

so you were thinking a doing flavor of this, 
with the previous suggested edit (removing lines from earlier in the code) ?

it kind of looks bad, because it's too long, but it should work,
Feel free to give another suggestion to make it prettier, I'm missing the way to prettify it, if that what you meant by removing _get_return_value_from_logs

```suggestion
                return self.cli.logs(container=self.container['Id'], stdout=True, stderr=True, stream=True) if self.xcom_all else self.cli.logs(container=self.container['Id'], stdout=True, stderr=True, stream=True, tail=1)
```

Comment (uranusjr) [reply airflow/providers/docker/operators/docker.py:335]: Yeah, something like this. We also need to deal with the `AirflowException` message above.

Comment (uranusjr) [reply airflow/providers/docker/operators/docker.py:335]: I wonder if `"".join(logstream).splitlines()` would achieve the same as the tail-less call.

Comment (asaf400) [reply airflow/providers/docker/operators/docker.py:335]: Maybe, I tested out adding random \n into one of the bash 'echo' (without -e) commands in my example in the issue discussion,
I tried using logstream.split('\n') but it had undesired results, as some of the random `\n`'s were converted to newlines in the output, but I forgot to test splitlines() I guess

Comment (asaf400) [reply airflow/providers/docker/operators/docker.py:335]: I'll test these suggestions shortly

Comment (asaf400) [reply airflow/providers/docker/operators/docker.py:335]: I guess this small test help understand what I meant

```python
#!/usr/bin/python3

str = r"""this is
an 'herestring'
with some lines,
but now\n there's even more \n lines, but are they real?\n
string example....\nwow!!!"""
print (str.splitlines())
```

In this case, the difference between an expected results and an unexpected results lays in the `r`,
without the r""", splitlines would also process the "inline `\n`'s"

I'll test how logstream would handle splitlines

File: airflow/providers/docker/operators/docker.py (hunk 1 above)
Comment (ferruzzi) [reply airflow/providers/docker/operators/docker.py:324]: @uranusjr  - I'm going to finish this one up.  Is there a reason we're pulling the logs again instead of reusing the parsed logs from above?

Comment (uranusjr) [reply airflow/providers/docker/operators/docker.py:324]: `logstream` only contains logs starting from the attach happens and may lose things between it and container startup. I’m not sure if it makes a difference, but honestly the additional API call is pretty cheap anyway, and might actually be more efficient for `xcom_all=False` because we don’t need to put all the previous logs into memory.

Comment (ferruzzi) [reply airflow/providers/docker/operators/docker.py:324]: Sure, but we're calling `attach` before we call `cli.start()` so that should mean the stream has everything from the beginning.  I'm happy to go either way, just thought it might be simplified while we were in there.

Comment (uranusjr) [reply airflow/providers/docker/operators/docker.py:324]: According to #9164, attching too slow and missing a line is a real problem though? I feel it’s just much safer to have a separate API call.
